ENH implement fair representation learner

taharallouche opened the pull request:
## Description
Tackles #1026

Hello ! Here are the most important details imo regarding this implementation:

### Comparison to original paper and AIF360 implementation

|  | Paper | AIF360 | FairRepresentationLearner |
|-------|-------|--------|--------------------------|
| Multi-class target     | :x:     | ✅ (*)    | :x: (**)                       |
| > 2 sensitive features groups     | ❌    | :x:      | ✅   (***)                    |
| sklearn-compatibility     | -    | ❌     | ✅                       |

-  (*) I'm not sure the implementation for multi-class target in AIF360 is correct though, the class probabilities `w` are bound 
to be in $(0,1)$ but they are not constrained to sum to $1$.
- (**) I think we can handle multi-class cases. The optimization problem will have linear constraints and would be solved with other optimizers than the `L-BFGS` one that the paper uses, among a couple of other things to adapt.
- (***) Achieved by including the absolute difference between each two groups in the $L_z$ loss term (more details in the user-guide).


### Fallback Estimator
In order to be sklearn-compatible, the `.fit` should work with the default `sensitive_features` kwarg value, which is `None`.
To achieve that, I made the choice to fallback to fitting a regular  `LogisticRegression` if no `sensitive_features` are passed.
The transformation in such cases will be the identity function, and the predictions are those of the fallback logistic regressor.


### Unit-tests
I have to say I feel the amount of unit tests is not much compared to the other modules, but I run out of ideas of code paths to cover, especially that the sklearn estimator checks already tests for many edge cases. 


## Tests
<!--- Select all that apply by putting an x between the brackets: [x] -->
- [ ] no new tests required
- [x] new tests added
- [ ] existing tests adjusted

## Documentation
<!--- Select all that apply. -->
- [ ] no documentation changes needed
- [x] user guide added or updated
- [x] API docs added or updated
- [ ] example notebook added or updated

## Screenshots
![image](https://github.com/user-attachments/assets/e3cabeed-9b4e-4139-b4aa-09b09085c369)
![image](https://github.com/user-attachments/assets/28f3b314-169c-478d-9031-42d2061cbf46)
![image](https://github.com/user-attachments/assets/d79b6722-b16a-4575-ba55-d26677f16f1d)
![image](https://github.com/user-attachments/assets/45ef657e-84fc-4215-889e-4353f116172a)

taharallouche commented:
I think that the CI is failing for the `windows 3.8` job because `pywinpty`'s [project.toml](https://github.com/andfoy/pywinpty/blob/main/pyproject.toml) is lacking `version` (`pywinpty` is a dep of `jupyterlite_sphinx`), here's a [related issue](https://github.com/andfoy/pywinpty/issues/486)

> project.version` field is required in pyproject.toml unless it is present in the `project.dynamic` list

`maturin` only [very recently](https://github.com/PyO3/maturin/pull/2391) started to require the field.

TamaraAtanasoska commented:
> I think that the CI is failing for the `windows 3.8` job because `pywinpty`'s [project.toml](https://github.com/andfoy/pywinpty/blob/main/pyproject.toml) is lacking `version` (`pywinpty` is a dep of `jupyterlite_sphinx`), here's a [related issue](https://github.com/andfoy/pywinpty/issues/486)
> 
> > project.version`field is required in pyproject.toml unless it is present in the`project.dynamic` list
> 
> `maturin` (which is not frozen in [pywinpty's build](https://github.com/andfoy/pywinpty/blob/main/pyproject.toml#L6)) only [very recently](https://github.com/PyO3/maturin/pull/2391) started to require the field.

I am back :) Wow that is a big PR thanks! I will look into the Windows issue, so we can separate it from the PR here and have a clean CI.

TamaraAtanasoska commented:
@taharallouche it seems that the error [was made less severe ](https://github.com/andfoy/pywinpty/issues/486#issuecomment-2572596171)so the builds pass now, and the library maintainers are keeping an eye on it. CI now builds with `pywinpty`, but there are unrelated test failures. There are some sklearn estimator checks failing around validation, try using [this function in the fixes](https://github.com/fairlearn/fairlearn/blob/main/fairlearn/utils/_fixes.py#L102) instead of the native one, as it is made to work with both versions before and after sklearn 1.6.

TamaraAtanasoska commented:
I get notifications so just popping here without a proper review to hopefully make your day easier. The sklearn 1.6 compatibility issues were working for what we had, trying to add minimal added code, but if you are using some other functions that we didn't have you might want to port some of this code found here https://github.com/sklearn-compat/sklearn-compat, and if it is too much and becomes annoying we can decide to vendor the whole library. It is a bit of a tedious thing, feel free to ping me on Discord about anything, I'd love to support if I can!

I wasn't clear enough the other day about the `parametrize_with_checks` compatibility function, you only need it if you have failed estimator checks that you want to place there, just for the easiness of development and having passing CI, the checks can be addressed later. Since the compatibility function only needs to be used when there are failing checks, the empty addition isn't necessary. I should add that to the docs as I realised it isn't there.

taharallouche commented:
Whoops it seems `scikit-learn 1.6.1` broke our `CorrelationRemover` and `AdversarialFairnessRegressor` 's compatibility :eyes: fyi @TamaraAtanasoska and @adrinjalali 

 (It's breaking my new estimator as well so I'll be looking into it ...)

TamaraAtanasoska commented:
hi @taharallouche, the main branch passes fully with `sklearn 1.6`, there are no issues, just rechecked again locally. it must be something in this one. I'd suggest that you remove the additions to the `validate_data` in _fixes.py you added earlier, as they are not necessary and only use the compatibility `parametrize_with_checks` if you have checks to add under the `xfail` flag.

TamaraAtanasoska commented:
hi @fairlearn/fairlearn-maintainers, this is now ready for a deeper review! I will sign myself up to do one next week, I have already looked into the branch a bit. A second reviewer is necessary. @hildeweerts as you have created the issue, would you be up for a review too?

hildeweerts commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+            The target values.
+
+        sensitive_features : array-like or None, default=None
+            Sensitive features to be considered whose groups will be used to enforce statistical
```
Enforce sounds perhaps a bit stronger than what the method does, perhaps something like improve/promote/increase/etc.?

hildeweerts reviewed (commented):
After a first pass, I’m unsure what a good API for this method should look like and/or how to categorize it properly. Intuitively, I wouldn’t expect a "pre-processing" algorithm to have a “predict” method (or need of a fall-back estimator).

According to the paper, the intuition behind the proposed representation learning approach is to pre-process data in an alternative representation, which could theoretically be reused by a different party for classification. Yet, the main experiments (section 4.2) use the probabilities derived from prototypes directly for classification, which seems more like an end-to-end representation learning approach rather than a pre-processing approach...

The chosen representation, which maps instances in the dataset probabilistically to prototypes, is a bit “weird” in the sense that it's not entirely clear how it would be used directly as input for a particular classification algorithm. In Section 4.3, they seem to only use the probabilities as a representation and not the learned prototypes (please correct me if I'm wrong).

I'm not sure if we need to change the implementation so much as the documentation. E.g. we might want to reconsider calling this a pre-processing method, make it clear in the API docs/user guide how the predicted probabilities are derived, show how intermediate transformations could be used in a pipeline with a different classification algorithm, etc.

Thoughts? @fairlearn/fairlearn-maintainers

TamaraAtanasoska commented:
before I start the review because I believe this PR will be affected too - I am working on a PR that will address the `codecov` issues https://github.com/fairlearn/fairlearn/pull/1448#issuecomment-2585236440 by adding a CI sklearn-compatibility job, as we codecov is complaining because we are not testing with both versions so all code isn't run https://github.com/fairlearn/fairlearn/issues/1485#issuecomment-2604633891.  I will make sure I have it proposed before I leave for the two weeks, but it might need to be merged before this one.

TamaraAtanasoska commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+    max_iter : int, default=1000
+        Maximum number of iterations for the optimization algorithm.
+
+    Attributes
```
`n_iter_`, `max_iter`, `_classes` and a few others are part of the necessary `scikit-learn` compatibility API, not sure if that speaks in favour or not. `scikit-learn` adds attributes to their docs, I can see how they make the code more accessible with that extra information bit available. Tools like Copilot or Cursor can now successfully add these automatically for you, update them too (of course you'd need to proofread), so it isn't a huge maintenance burden. I personally am ok with either way.

TamaraAtanasoska reviewed (commented):
Thank you again for all the hard work! Since we cleared out most of the sklearn issues before, I have general stylistic nitpicks and a few questions. I want to think a bit if we could test something else too, but we can also add more tests in a subsequent PR as we discuss reproducibility further for all the methods.

taharallouche commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+    random_state : int, np.random.RandomState, or None, default=None
+        Seed or random number generator for reproducibility.
+
+    optimizer : Literal["L-BFGS-B", "Nelder-Mead", "Powell", "SLSQP", "TNC", "trust-constr",
```
TBH, the paper only mentions `L-BFGS`, the others are the algorithms that can solve this kind of minimization with bounds constraints among the list of algorithms that `scipy` offers. That is to say, I don't have any recommendation, except sticking to L-BFGS if one wants to mimick the paper's approach :pensive: 

@TamaraAtanasoska I think this answers your question below too.

TamaraAtanasoska replied:
I see! I am not sure what the concrete impact would be on the calculations, but I guess we can see this two ways: if we stick to academic reproducibility rules we should stick to what is in the paper, but since we aren't doing this as part of an academic exercise we also have some space for own extensions as long as we document them with our reasoning for it (which can also be "other algorithms that can solve this kind of minimization with bounds constraints among the list of algorithms that scipy offers" or something along the lines, with a disclaimer that if one wants to be true to the paper should keep the one there which I guess should be the default too).

taharallouche replied:
I'm beginning to think that it wasn't actually a great idea to add these options without having a proper motivation or a well justified reason to do so ..
If something goes wrong with an algorithm other than L-BFGS it would be our responsibility.

I'm now more inclined to remove them.

TamaraAtanasoska replied:
we can always add them in the future, or respond to user needs if they want some added.

MiroDudik replied:
Yeah--I think we should just stick with one optimization algorithm, so we don't need the optimizer argument.

taharallouche commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+    n_prototypes : int, default=2
+        Number of prototypes to use in the latent representation.
+
+    Ax : float, default=1.0
```
Since alpha is already an optimization parameter (the weight of each dimension in the  distance) I used `lamnda_x`, `lambda_y,` `lambda_z`

taharallouche commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+            expect_sensitive_features=False,
+            enforce_binary_labels=True,
+        )
+        assert sensitive_features is None or isinstance(sensitive_features, pd.Series)
```
It doesn't have to be a series per se, but the `_validate_and_reformat_input` converts the user input (which can be any array-like type) to a series, which is then used downstream

TamaraAtanasoska replied:
thanks for the clarification!

taharallouche commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+        )
+
+        def objective(x: np.ndarray, X, y) -> float:
+            assert x.shape == (self._optimizer_size,)
```
This one, in contrast with the first assert, I'd prefer to keep (no strong opinion though), since it's safeguarding against [a corrupt program state](https://stackoverflow.com/a/945135) which shouldn't happen given that we set the initial point ourselves and run the algorithm

TamaraAtanasoska replied:
I see! In the case of never happening it makes sense. Resolving :)

adrinjalali reviewed (commented):
We have classifiers which also transform, so that's nothing new, but calling it a preprocessing step is odd.

TamaraAtanasoska commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+                options={"maxiter": self.max_iter},
+            )
+        except Exception as optimization_error:
+            raise RuntimeError("The loss minimization failed.") from optimization_error
```
Ah fair, I understand. I don't have a better idea at the moment what to say there, as a generic addition like "Consider changing some of the parameters or inputs" brings absolutely no value and most likely the folks running it will have the experience to figure that out. I am ok with leaving it like this!

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -139,3 +139,100 @@ we would expect with :math:`\alpha=0.5`.
+
+.. _fair_representation_learner:
+
+Fair Representation Learner
```
In the past, we have focused on the descriptive aspects of the approach (like "ThresholdOptimizer" and "CorrelationRemover"), so one thought would be to talk about "PrototypedRepresenter" or something like that?

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -139,3 +139,100 @@ we would expect with :math:`\alpha=0.5`.
+
+.. math::
+
+    L_z = \frac{1}{K \times \binom{G}{2}} \sum_{k=1}^{K} \sum_{g < g'} \left| M_{gk} - M_{g'k} \right|
```
There are other options here that would be more compatible with how we do things elsewhere. For example, we could use
\sum_{g=1}^G \sum_{k=1}^K | M_{gk} - M_{*k} |
where M_{*k} =\frac{1}{N} \sum_{n=1}^n M_{nk}. This would however not be compatible with the paper for G=2.

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -139,3 +139,100 @@ we would expect with :math:`\alpha=0.5`.
+The latent representation consists of :math:`K` prototype vectors
+:math:`\mathbf{v}_1, \dots, \mathbf{v}_K \in \mathbb{R}^d`, where :math:`d` is the dimension
+of the input data, and a stochastic transformation matrix :math:`\mathbf{M} \in \mathbb{R}^{n \times k}`
+that maps the input data :math:`\mathbf{X} \in \mathbb{R}^{n \times d}` to the latent representation.
```
Based on reading the paper, there are several ways to transform the data (they are all based on solving the same optimization problem):

1. x_n -> [M_n1,...,M_nK]
   * this is what the authors seem to assume in Section 4.2 (Eq. 16) and Section 4.3
2. x_n -> \hat{x}_n
   * this nominally uses the same features as the original data, but it can turn 0/1 features into real-valued features 
3. x_n -> sample an index k_n\in{1,...,K} according to weights M_n1,...,M_nK, and return v_{k_n}
   * this is what actually achieves the goal that authors claim (about the mutual information between S and Z) also, it seems to be most in line with the logic in Section 2.3
   * the approach can turn 0/1 features into real-valued features, but here it is not a fundamental property of the approach--we could possibly use a different optimizer, which looks for prototypes that satisfy some additional constraints (like 0/1 constraints on the feature values, or prototypes would need to correspond to some examples from the dataset, etc.)

taharallouche replied:
Thinking about it, I think that 1 and 3 are the ones most inline with the goal of the paper; obfuscating the sensitive information.

We could add a kwarg to `.transform`, to allow the user to choose the transformation type (which would default to 1, to match the paper's methodology)

MiroDudik replied:
I chatted with @hildeweerts, and we thought that implementing (1) might be the most obvious choice since that's what the paper seems to be doing. In that case, we would still put options 2 & 3 in the documentation.

However, when I looked at [AIF360 implementation](https://github.com/Trusted-AI/AIF360/blob/main/aif360/sklearn/preprocessing/learning_fair_representations.py), they seem to do option (2).

As a result, I'm ambivalent between implementing (1) or implementing (1,2,3) but still defaulting to (1). I'd probably go for just implementing (1) for simplicity, but I can also see arguments for implementing (1,2,3).

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -139,3 +139,100 @@ we would expect with :math:`\alpha=0.5`.
+
+    L_x = \frac{1}{N} \sum_{n=1}^{N} \| \mathbf{x}_n - \hat{\mathbf{x}}_n \|_2^2
+
+:math:`L_y` is the classification error term that is equal to the log loss:
```
I think that the approach makes sense even if there are no labels, in which case the objective L_y would be missing.

taharallouche replied:
We do support that by passing zero weight `lambda_y = 0` to the classification error term in the loss.

Would you recommend removing the `Classifier` aspect of the class, and only maintaining it as a `Transformer` ? (it would spare me the `fallback_estimator` hack , among other things). @hildeweerts raises [a similar point](https://github.com/fairlearn/fairlearn/pull/1478#pullrequestreview-2556322881) if I understood correctly

MiroDudik replied:
I am fine with just implementing `TransformerMixin` or both `TransformerMixin` and `ClassifierMixin`. However, if we implement the classifier, we should get rid of the fallback classifier and simply throw an exception if no sensitive features are provided. (Alternatively, we could still run optimization but without the fairness term; that would basically correspond to a variant of a nearest-neighbor classifier.)

For what it's worth: AIF360 is implementing the classifier interface; the classifier was also included in the original paper, so I think it would be okay and quite natural to include it.

taharallouche commented:
@MiroDudik I won't be able to attend today's meeting, if you have some time ,could you please take a look at these points so that I can move forward with this one ? :pray: 

- https://github.com/fairlearn/fairlearn/pull/1478#discussion_r1957170668 (Defining the L_z term in the loss)
- https://github.com/fairlearn/fairlearn/pull/1478#discussion_r1957172171 (Defining the transformation)
- https://github.com/fairlearn/fairlearn/pull/1478#discussion_r1957171744 (Dropping the Classifier aspect of the estimator)

MiroDudik commented:
@taharallouche : hopefully you're mostly unblocked. In addition to the above, a couple more requests:
* let's call the class `PrototypeRepresentationLearner`. It's a bit verbose, but since the paper is about "learning fair representations", this name will sound closer to that in spirit
* for L_x, L_y, L_z we can maybe just use the same terminology as AIF360, which is much more mnemonic: reconstruct_weight, target_weight, fairness_weight

taharallouche commented on fairlearn/preprocessing/_fair_representation_learner.py:
```diff
@@ -0,0 +1,621 @@
+    tol : float, default=1e-6
+        Convergence tolerance for the optimization algorithm.
+
+    max_iter : int, default=1000
```
I'm still looking for resources (regarding the convergence of the L-BFGS-B algorithm) to base our recommendation on ...

taharallouche commented:
> @taharallouche : hopefully you're mostly unblocked. In addition to the above, a couple more requests:
> 
>     * let's call the class `PrototypeRepresentationLearner`. It's a bit verbose, but since the paper is about "learning fair representations", this name will sound closer to that in spirit
> 
>     * for L_x, L_y, L_z we can maybe just use the same terminology as AIF360, which is much more mnemonic: reconstruct_weight, target_weight, fairness_weight

@MiroDudik thank you ! I believe I addressed all of the points above.

taharallouche commented:
One question that is still up for discussion, is whether we want to define/present this as a pre-processing algorithm (see @hildeweerts 's [comment](https://github.com/fairlearn/fairlearn/pull/1478#pullrequestreview-2556322881)).

I personally am ok with that, especially now that the `.fit` / `transform`  works even in absence of targets `y`.

hildeweerts commented:
> One question that is still up for discussion, is whether we want to define/present this as a pre-processing algorithm (see @hildeweerts 's [comment](https://github.com/fairlearn/fairlearn/pull/1478#pullrequestreview-2556322881)).
> 
> I personally am ok with that, especially now that the `.fit` / `transform` works even in absence of targets `y`.

I would also be okay with that, @taharallouche!

TamaraAtanasoska commented:
This seems ready to merge to me unless there are any objections. @taharallouche would you agree?

TamaraAtanasoska commented:
@taharallouche me and @hildeweerts had a short chat yesterday about it and I looked over the comments - it seems everything is answered and in a good shape, I believe anyone that wanted to chime in did so. I really like the `PrototypeRepresentationLearner` name. It would be nice to create an example notebook about this method, we could make an issue about it following the paper in the lack of a more real world use case.
@MiroDudik do you want to throw a last look, or we are good to go?

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -139,3 +139,117 @@ we would expect with :math:`\alpha=0.5`.
+
+.. _prototyped_representation_learner:
+
+Prototyped Representation Learner
```
```suggestion
Prototype Representation Learner
```

taharallouche commented:
@TamaraAtanasoska I addressed Miro's last comments, good to go ? :rocket:
